File: ButtonBackgroundColorTests/ButtonBackgroundColorTests.m
//
//  ButtonBackgroundColorTests.m
//  ButtonBackgroundColorTests
//
//  Created by GabrielMassana on 04/03/2016.
//  Copyright © 2016 GabrielMassana. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "UIButton+BBCBackgroundColor.h"

@interface ButtonBackgroundColorTests : XCTestCase

@property (nonatomic, strong) UIButton *button;
@property (nonatomic, strong) UIButton *deprecatedButton;
@property (nonatomic, strong) UIColor *normalColor;
@property (nonatomic, strong) UIColor *highlightedColor;

@end

@implementation ButtonBackgroundColorTests

- (void)setUp
{
    [super setUp];
    
    self.normalColor = [UIColor redColor];
    self.highlightedColor = [UIColor orangeColor];
    
    self.button = [UIButton buttonWithType:UIButtonTypeCustom];
    
    [self.button bbc_backgroundColorNormal:self.normalColor
                backgroundColorHighlighted:self.highlightedColor];
    
    self.deprecatedButton = [UIButton buttonWithType:UIButtonTypeCustom];
    
    [self.deprecatedButton bbc_backgroundColorNormal:self.normalColor
                backgroundColorHighlighted:self.highlightedColor];
}

- (void)tearDown
{
    self.normalColor = nil;
    self.highlightedColor = nil;
    self.button = nil;
    self.deprecatedButton = nil;
    
    [super tearDown];
}

#pragma mark - Color

- (void)test_bbc_backgroundColor_setted
{
    XCTAssertEqualObjects(self.button.backgroundColor, self.normalColor,  @"fail_test_bbc_backgroundColor_setted");
}

- (void)test_bbc_normalColor_stored
{
    XCTAssertEqualObjects(self.button.normalBackgroundColor, self.normalColor,  @"fail_test_bbc_normalColor_stored");
}

- (void)test_bbc_highlightedColor_stored
{
    XCTAssertEqualObjects(self.button.highlightedBackgroundColor, self.highlightedColor,  @"fail_test_bbc_highlightedColor_stored");
}

#pragma mark - ColorDeprecated

- (void)test_bbc_backgroundColor_setted_deprecated
{
    XCTAssertEqualObjects(self.deprecatedButton.backgroundColor, self.normalColor,  @"fail_test_bbc_backgroundColor_setted_deprecated");
}

- (void)test_bbc_normalColor_stored_deprecated
{
    XCTAssertEqualObjects(self.deprecatedButton.normalBackgroundColor, self.normalColor,  @"fail_test_bbc_normalColor_stored_deprecated");
}

- (void)test_bbc_highlightedColor_stored_deprecated
{
    XCTAssertEqualObjects(self.deprecatedButton.highlightedBackgroundColor, self.highlightedColor,  @"fail_test_bbc_highlightedColor_stored_deprecated");
}

#pragma mark - ControlEvent

- (void)test_bbc_controlEventTouchDown
{
    [self.button sendActionsForControlEvents:UIControlEventTouchDown];
    
    XCTAssertEqualObjects(self.button.backgroundColor, self.highlightedColor,  @"fail_test_bbc_controlEventTouchDown");
}

- (void)test_bbc_controlEventTouchUpInside
{
    [self.button sendActionsForControlEvents:UIControlEventTouchUpInside];
    
    XCTAssertEqualObjects(self.button.backgroundColor, self.normalColor,  @"fail_test_bbc_controlEventTouchUpInside");
}

- (void)test_bbc_controlEventTouchUpOutside
{
    [self.button sendActionsForControlEvents:UIControlEventTouchUpOutside];
    
    XCTAssertEqualObjects(self.button.backgroundColor, self.normalColor,  @"fail_test_bbc_controlEventTouchUpOutside");
}

- (void)test_bbc_controlEventTouchCancel
{
    [self.button sendActionsForControlEvents:UIControlEventTouchCancel];
    
    XCTAssertEqualObjects(self.button.backgroundColor, self.normalColor,  @"fail_test_bbc_controlEventTouchCancel");
}

#pragma mark - ControlEventDeprecated

- (void)test_bbc_controlEventTouchDown_deprecated
{
    [self.deprecatedButton sendActionsForControlEvents:UIControlEventTouchDown];
    
    XCTAssertEqualObjects(self.deprecatedButton.backgroundColor, self.highlightedColor,  @"fail_test_bbc_controlEventTouchDown_deprecated");
}

- (void)test_bbc_controlEventTouchUpInside_deprecated
{
    [self.button sendActionsForControlEvents:UIControlEventTouchUpInside];
    
    XCTAssertEqualObjects(self.deprecatedButton.backgroundColor, self.normalColor,  @"fail_test_bbc_controlEventTouchUpInside_deprecated");
}

- (void)test_bbc_controlEventTouchUpOutside_deprecated
{
    [self.button sendActionsForControlEvents:UIControlEventTouchUpOutside];
    
    XCTAssertEqualObjects(self.deprecatedButton.backgroundColor, self.normalColor,  @"fail_test_bbc_controlEventTouchUpOutside_deprecated");
}

- (void)test_bbc_controlEventTouchCancel_deprecated
{
    [self.button sendActionsForControlEvents:UIControlEventTouchCancel];
    
    XCTAssertEqualObjects(self.deprecatedButton.backgroundColor, self.normalColor,  @"fail_test_bbc_controlEventTouchCancel_deprecated");
}

#pragma mark - ControlState

- (void)test_bbc_controlStateNormal
{
    XCTAssertEqualObjects(self.button.backgroundColor, self.normalColor,  @"fail_test_bbc_controlStateNormal");
}

- (void)test_bbc_controlStateHighlighted
{
    self.button.highlighted = YES;
    
    XCTAssertEqualObjects(self.button.backgroundColor, self.highlightedColor,  @"fail_test_bbc_controlStateHighlighted");
}

#pragma mark - ControlStateDeprecated

- (void)test_bbc_controlStateNormal_deprecated
{
    XCTAssertEqualObjects(self.deprecatedButton.backgroundColor, self.normalColor,  @"fail_test_bbc_controlStateNormal_deprecated");
}

- (void)test_bbc_controlStateHighlighted_deprecated
{
    self.deprecatedButton.highlighted = YES;
    
    XCTAssertEqualObjects(self.deprecatedButton.backgroundColor, self.highlightedColor,  @"fail_test_bbc_controlStateHighlighted_deprecated");
}

@end

File: ButtonBackgroundColorTests/ButtonBackgroundColorTests.swift
//
//  ButtonBackgroundColorTests.swift
//  ButtonBackgroundColorTests
//
//  Created by GabrielMassana on 12/04/2016.
//  Copyright © 2016 GabrielMassana. All rights reserved.
//

import XCTest

@testable import ButtonBackgroundColor

class ButtonBackgroundColorTests: XCTestCase {
    
    //MARK: - Accessors

    var button: UIButton?
    var buttonWithoutExtension: UIButton?
    
    var normalColor: UIColor?
    var highlightedColor: UIColor?
    
    //MARK: - SetUp

    override func setUp()
    {
        super.setUp()
        
        normalColor = UIColor.red
        highlightedColor = UIColor.orange
    
        button = UIButton(type: .custom)
        
        button!.backgroundColorForStates(normal: normalColor!, highlighted: highlightedColor!)
        
        buttonWithoutExtension = UIButton(type: .custom)
        
        buttonWithoutExtension!.backgroundColor = normalColor
    }
    
    override func tearDown()
    {
        self.normalColor = nil;
        self.highlightedColor = nil;
        
        self.button = nil;
        self.buttonWithoutExtension = nil;
        
        super.tearDown()
    }
    
    //MARK: - Color

    func test_bbc_backgroundColor_setted()
    {
        XCTAssertEqual(button!.backgroundColor, normalColor,  "fail_test_bbc_backgroundColor_setted");
    }
    
    func test_bbc_normalColor_stored()
    {
        XCTAssertEqual(button!.normalBackgroundColor(), normalColor,  "fail_test_bbc_normalColor_stored");
    }
    
    func test_bbc_highlightedColor_stored()
    {
        XCTAssertEqual(button!.highlightedBackgroundColor(), highlightedColor,  "fail_test_bbc_highlightedColor_stored");
    }
    
    //MARK: - ColorWithoutExtension

    func test_bbc_backgroundColor_setted_withoutExtension()
    {
        XCTAssertEqual(buttonWithoutExtension!.backgroundColor, normalColor,  "fail_test_bbc_backgroundColor_setted_withoutExtension");
    }
    
    func test_bbc_normalColor_stored_withoutExtension()
    {
        XCTAssertEqual(buttonWithoutExtension!.normalBackgroundColor(), nil,  "fail_test_bbc_normalColor_stored_withoutExtension");
    }
    
    func test_bbc_highlightedColor_stored_withoutExtension()
    {
        XCTAssertEqual(buttonWithoutExtension!.highlightedBackgroundColor(), nil,  "fail_test_bbc_highlightedColor_stored_withoutExtension");
    }
    
    //MARK: - ControlEvent
    
    func test_bbc_controlEventTouchDown()
    {
        button!.sendActions(for: .touchDown)
        
        XCTAssertEqual(button!.backgroundColor, highlightedColor,  "fail_test_bbc_controlEventTouchDown");
    }
    
    func test_bbc_controlEventTouchUpInside()
    {
        button!.sendActions(for: .touchUpInside)
        
        XCTAssertEqual(button!.backgroundColor, normalColor,  "test_bbc_controlEventTouchUpInside");
    }
    
    func test_bbc_controlEventTouchUpOutside()
    {
        button!.sendActions(for: .touchUpOutside)
        
        XCTAssertEqual(button!.backgroundColor, normalColor,  "test_bbc_controlEventTouchUpOutside");
    }
    
    func test_bbc_controlEventTouchCancel()
    {
        button!.sendActions(for: .touchCancel)
        
        XCTAssertEqual(button!.backgroundColor, normalColor,  "test_bbc_controlEventTouchCancel");
    }
    
    //MARK: - ControlEventWithoutExtension
    
    func test_bbc_controlEventTouchDown_withoutExtension()
    {
        buttonWithoutExtension!.sendActions(for: .touchDown)
        
        XCTAssertEqual(button!.backgroundColor, normalColor,  "test_bbc_controlEventTouchDown_withoutExtension");
    }
    
    func test_bbc_controlEventTouchUpInside_withoutExtension()
    {
        buttonWithoutExtension!.sendActions(for: .touchUpInside)
        
        XCTAssertEqual(button!.backgroundColor, normalColor,  "test_bbc_controlEventTouchUpInside_withoutExtension");
    }
    
    func test_bbc_controlEventTouchUpOutside_withoutExtension()
    {
        buttonWithoutExtension!.sendActions(for: .touchUpOutside)
        
        XCTAssertEqual(button!.backgroundColor, normalColor,  "test_bbc_controlEventTouchUpOutside_withoutExtension");
    }
    
    func test_bbc_controlEventTouchCancel_withoutExtension()
    {
        buttonWithoutExtension!.sendActions(for: .touchCancel)
        
        XCTAssertEqual(button!.backgroundColor, normalColor,  "test_bbc_controlEventTouchCancel_withoutExtension");
    }
}
